File: src/main/java/miniDB/index/bptree/BPlusConfiguration.java
package miniDB.index.bptree;


import com.lss233.minidb.engine.storage.StorageType;
import com.lss233.minidb.exception.MiniDBException;

import java.util.ArrayList;

/**
 * Class that stores all the configuration parameters for our B+ Tree.
 * You can view a description on all the parameters below...
 */
@SuppressWarnings({"WeakerAccess", "unused"})
public class BPlusConfiguration extends Configuration {

    public int valueSize;          // entry size (in bytes)
    public int headerSize;               // header size (in bytes)

    public int leafHeaderSize;           // leaf node header size (in bytes)
    public int internalNodeHeaderSize;   // internal node header size (in bytes)
    public int overflowNodeHeaderSize;   // overflow node header size
    public int freePoolNodeHeaderSize; // free pool page header size

    public int leafNodeDegree;           // leaf node degree
    public int treeDegree;               // tree degree (internal node degree)
    public int overflowPageDegree;       // overflow page degree
    public int freePoolNodeDegree; // lookup overflow page degree

    public int trimFileThreshold;       // iterations to trim the file
    public boolean unique;               // whether one key can have multiple values. This corresponds to unique index.

    public BPlusConfiguration(int pageSize, int valueSize, ArrayList<StorageType> types, ArrayList<Integer> sizes, ArrayList<Integer> colIDs,
                              boolean unique, int trimFileThreshold)
            throws MiniDBException {
        super(pageSize, types, sizes, colIDs);
        this.unique = unique;
        this.valueSize = valueSize; // entry size (in bytes)
        this.trimFileThreshold = trimFileThreshold;       // iterations for conditioning

        this.headerSize = (Integer.SIZE * 3 + 4 * Long.SIZE) / 8;          // header size in bytes

        this.leafHeaderSize = (Short.SIZE + 2 * Long.SIZE + Integer.SIZE) / 8; // 22 bytes
        this.internalNodeHeaderSize = (Short.SIZE + Integer.SIZE) / 8; // 6 bytes
        this.overflowNodeHeaderSize = (Short.SIZE + 2 * Long.SIZE + Integer.SIZE) / 8 + keySize; // 22 + keySize bytes
        this.freePoolNodeHeaderSize = (Short.SIZE + Long.SIZE + Integer.SIZE) / 8; // 14 bytes

        // now calculate the degree

        // data: key and a value and an overflow pointer
        this.leafNodeDegree = calculateDegree(keySize + valueSize + Long.SIZE / 8, leafHeaderSize);
        // data: key and a pointer
        this.treeDegree = (pageSize - internalNodeHeaderSize - Long.SIZE / 8) / (keySize + Long.SIZE / 8);
        this.overflowPageDegree = calculateDegree(valueSize, overflowNodeHeaderSize);
        this.freePoolNodeDegree = calculateDegree(Long.SIZE / 8, freePoolNodeHeaderSize);
        checkDegreeValidity();
    }

    private int calculateDegree(int elementSize, int elementHeaderSize) {
        return (pageSize-elementHeaderSize)/elementSize;
    }

    /**
     * Little function that checks if we have any degree < 2 (which is not allowed)
     */
    private void checkDegreeValidity() {
        if (treeDegree < 2 || leafNodeDegree < 2 || overflowPageDegree < 2 || freePoolNodeDegree < 2) {
            throw new IllegalArgumentException("Can't have a degree < 2");
        }
    }

    public int getMaxInternalNodeCapacity() {
        return treeDegree;
    }

    public int getMinInternalNodeCapacity() {
        return (treeDegree-1) / 2;
    }

    public int getMaxLeafNodeCapacity() {
        return leafNodeDegree;
    }

    public int getMinLeafNodeCapacity() {
        return (leafNodeDegree-1) / 2;
    }

    public int getMaxOverflowNodeCapacity() {
        return overflowPageDegree;
    }

    public int getFreePoolNodeDegree() {
        return freePoolNodeDegree;
    }

    public long getFreePoolNodeOffset() {
        return freePoolNodeHeaderSize;
    }

    public int getPageCountOffset() {
        return 12;
    }

}

File: src/main/kotlin/com/lss233/minidb/engine/index/bptree/BPlusConfiguration.kt
package com.lss233.minidb.engine.index.bptree

import java.lang.reflect.Type

/**
 * Class that stores all the configuration parameters for our B+ Tree.
 * You can view a description on all the parameters below...
 * @param valueSize entry size (in bytes)
 * @param unique whether one key can have multiple values. This corresponds to unique index.
 * @param trimFileThreshold iterations to trim the file
 */
@SuppressWarnings("WeakerAccess", "unused")
class BPlusConfiguration(
    pageSize: Int,
    var valueSize: Int = 0,
    types: ArrayList<Type>,
    sizes: ArrayList<Int>,
    colIDs: ArrayList<Int>,
    var unique: Boolean = false,
    var trimFileThreshold: Int = 0
): Configuration(pageSize, types, sizes,colIDs) {

    var headerSize: Long = 0 // header size (in bytes)

    private var leafHeaderSize:Long = 0 // leaf node header size (in bytes)

    private var internalNodeHeaderSize:Long = 0 // internal node header size (in bytes)

    private var overflowNodeHeaderSize:Long = 0 // overflow node header size

    private var freePoolNodeHeaderSize:Long = 0 // free pool page header size


    private var leafNodeDegree:Long = 0 // leaf node degree

    private var treeDegree:Long = 0 // tree degree (internal node degree)

    private var overflowPageDegree:Long = 0 // overflow page degree

    var freePoolNodeDegree:Long = 0 // lookup overflow page degree


    init {

        headerSize = (((Int.SIZE_BITS * 3 + 4 * Long.SIZE_BITS) / 8).toLong()) // header size in bytes

        leafHeaderSize = ((Short.SIZE_BITS + 2 * Long.SIZE_BITS + Int.SIZE_BITS) / 8).toLong() // 22 bytes

        internalNodeHeaderSize = ((Short.SIZE_BITS + Int.SIZE_BITS) / 8).toLong() // 6 bytes

        overflowNodeHeaderSize =
            ((Short.SIZE_BITS + 2 * Long.SIZE_BITS + Int.SIZE_BITS) / 8 + keySize).toLong() // 22 + keySize bytes

        freePoolNodeHeaderSize = ((Short.SIZE_BITS + Long.SIZE_BITS + Int.SIZE_BITS) / 8).toLong() // 14 bytes

        // now calculate the degree

        // data: key and a value and an overflow pointer
        leafNodeDegree = calculateDegree(keySize + valueSize + Long.SIZE_BITS / 8, leafHeaderSize)
        // data: key and a pointer
        treeDegree = (pageSize - internalNodeHeaderSize - Long.SIZE_BITS / 8) / (keySize + Long.SIZE_BITS / 8)

        overflowPageDegree = calculateDegree(valueSize, overflowNodeHeaderSize)

        freePoolNodeDegree = calculateDegree(Long.SIZE_BITS / 8, freePoolNodeHeaderSize)

        checkDegreeValidity()
    }

    private fun calculateDegree(elementSize: Int, elementHeaderSize: Long): Long {
        return (pageSize - elementHeaderSize) / elementSize
    }


    /**
     * Little function that checks if we have any degree < 2 (which is not allowed)
     */
    private fun checkDegreeValidity() {
        require(!(treeDegree < 2 || leafNodeDegree < 2 || overflowPageDegree < 2 || freePoolNodeDegree < 2)) { "Can't have a degree < 2" }
    }

    fun getMaxInternalNodeCapacity(): Long {
        return treeDegree
    }

    fun getMinInternalNodeCapacity(): Long {
        return (treeDegree - 1) / 2
    }

    fun getMaxLeafNodeCapacity(): Long {
        return leafNodeDegree
    }

    fun getMinLeafNodeCapacity(): Long {
        return (leafNodeDegree - 1) / 2
    }

    fun getMaxOverflowNodeCapacity(): Long {
        return overflowPageDegree
    }

    fun getFreePoolNodeOffset(): Long {
        return freePoolNodeHeaderSize
    }

    fun getPageCountOffset(): Long {
        return 12
    }
}
